### PR TITLE
Added support for bitwise AND, OR, XOR and NOT ops.

### PR DESCRIPTION
**Pull request by kmitrovicTT**
- Defined ops in TTIR and TTNN dialects.
- Implemented StableHLO to TTIR converison (unified with logical ops)
- Implemented TTIR to TTNN conversion
- Added tests

Fixes https://github.com/tenstorrent/tt-mlir/issues/1202.

Half of solution (tt-xla and tt-torch tests are second half) for issues:
https://github.com/tenstorrent/tt-mlir/issues/1051
https://github.com/tenstorrent/tt-mlir/issues/1053
https://github.com/tenstorrent/tt-mlir/issues/1054
https://github.com/tenstorrent/tt-mlir/issues/1055

Left asserts in runtime code due to https://github.com/tenstorrent/tt-metal/issues/13582.

### CODE REVIEW

**Review by sdjordjevicTT — APPROVED** [review]

File: include/ttmlir/Target/TTNN/program.fbs [hunk 1]
```diff
@@ -94,24 +94,28 @@ enum EltwiseOpType: uint32 {
...
-  Gelu = 36,
-  LogicalXor = 37,
-  Clamp = 38,
-  LeakyRelu = 39,
+  LogicalXor = 22,
+  LogicalNot = 23,
+  BitwiseAnd = 24,
+  BitwiseOr = 25,
+  BitwiseXor = 26,
+  BitwiseNot = 27,
```
Comment (mrakitaTT) [inline include/ttmlir/Target/TTNN/program.fbs:102]: Don't forget to remove explicit enum values after rebase

File: test/ttmlir/Silicon/TTNN/simple_bitwise.mlir
Comment (mrakitaTT) [inline test/ttmlir/Silicon/TTNN/simple_bitwise.mlir:1]: I'd split this file into bitwise binary and unary, and then mark the binary one with `// UNSUPPORTED: true` since it is expected to fail in runtime

Comment (mrakitaTT) [inline test/ttmlir/Silicon/TTNN/simple_bitwise.mlir:1]: You also probably need to add `perf_unit` tests for the new ops

**Review by mrakitaTT — APPROVED** [review]

**Comment by kmitrovicTT** [comment]
Raised issue https://github.com/tenstorrent/tt-mlir/issues/1571 which causes `bitwise_not` to fail.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 2]
```diff
@@ -1948,20 +1960,21 @@ void addCCLOpsConversionPattern(MLIRContext *ctx, RewritePatternSet &patterns,
...
-                                                             ctx);
-  patterns.add<StableHLOToTTIROpLogicalOpConversionPattern<
-      mlir::stablehlo::NotOp, mlir::tt::ttir::LogicalNotOp>>(typeConverter,
-                                                             ctx);
-  patterns.add<StableHLOToTTIROpLogicalOpConversionPattern<
-      mlir::stablehlo::OrOp, mlir::tt::ttir::LogicalOrOp>>(typeConverter, ctx);
-  patterns.add<StableHLOToTTIROpLogicalOpConversionPattern<
-      mlir::stablehlo::XorOp, mlir::tt::ttir::LogicalXorOp>>(typeConverter,
-                                                             ctx);
+void addLogicalAndBitwiseOpsConversionPatterns(MLIRContext *ctx,
```
Comment (mmanzoorTT) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:1963]: Please add a comment to explain two Dest ops either here or with the class `StableHLOToTTIRLogicalAndBitwiseOpConversionPattern`. It is bit different than the other patterns. So it will be helpful to have some explanation.

File: lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp [hunk 3]
```diff
@@ -666,11 +667,14 @@ void populateTTNNToEmitCPatterns(mlir::MLIRContext *ctx,
   // Eltwise binary ops
   //
   patterns.add<EltwiseBinaryOpConversionPattern<ttnn::AddOp>,
+               EltwiseBinaryOpConversionPattern<ttnn::SubtractOp>,
```
Comment (mmanzoorTT) [inline lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp:670]: Please revert these changes (line 670-671).

File: test/ttmlir/Silicon/TTNN/simple_bitwise_unary.mlir [hunk 4]
```diff
@@ -0,0 +1,17 @@
+// RUN: ttmlir-opt --ttir-to-ttnn-backend-pipeline="system-desc-path=%system_desc_path%" %s > %t.mlir
+// RUN: FileCheck %s --input-file=%t.mlir
+// RUN: ttmlir-translate --ttnn-to-flatbuffer %t.mlir > %t.ttnn
+// TODO(kmitrovic): Failing due to #1571
```
Comment (mmanzoorTT) [inline test/ttmlir/Silicon/TTNN/simple_bitwise_unary.mlir:4]: Please add complete link instead of `#1571`

**Review by mrakitaTT — APPROVED** [review]

**Review by mmanzoorTT — APPROVED** [review]

File: include/ttmlir/Dialect/TTNN/IR/TTNNOps.td [hunk 5]
```diff
@@ -255,6 +255,19 @@ def TTNN_LogicalNotOp: TTNN_ElementwiseUnaryOp<"logical_not"> {
     }];
 }
 
+def TTNN_BitwiseNotOp : TTNN_ElementwiseUnaryOp<"bitwise_not"> {
```
Comment (svuckovicTT) [inline include/ttmlir/Dialect/TTNN/IR/TTNNOps.td:258]:
```c
    static Tensor invoke(
        const Tensor& input_tensor,
        const std::optional<MemoryConfig>& memory_config = std::nullopt,
        const std::optional<Tensor>& optional_output_tensor = std::nullopt);
```

This op has an optional memory config operand as well, we should add it here. All the other ops that were added should have their operands match signatures from ttnn lib.

File: lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp (hunk 3 above)
Comment (kmitrovicTT) [reply lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp:670]: I tried putting bitwise and logical ops next to each other in all parts of code, just to have cleaner grouping. Same here, grouping on the right seemed better than the one on the left.

image

If you still prefer the former way, I'll revert the changes.

File: include/ttmlir/Dialect/TTNN/IR/TTNNOps.td (hunk 5 above)
Comment (kmitrovicTT) [reply include/ttmlir/Dialect/TTNN/IR/TTNNOps.td:258]: In an offline agreement, we decided to raise separate [issue](https://github.com/tenstorrent/tt-mlir/issues/1612) for this, and leave things as is in this PR.

File: lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp (hunk 3 above)
Comment (svuckovicTT) [reply lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp:670]: I prefer it with Kristijan's changes, the grouping makes more sense, let's leave it like that.

**Review by svuckovicTT — APPROVED** [review]

File: lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp (hunk 3 above)
Comment (kmitrovicTT) [reply lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp:670]: Asif agreed in an online discussion.

**Review by AleksKnezevic — APPROVED** [review]
